ui/packages/consul-ui/app/components/data-writer/index.js

```diff
@@ -21,5 +21,17 @@ export default Component.extend(Slotted, {
       set(this, 'data', data);
       this.dispatch('PERSIST');
     },
+    error: function(data, e) {
+      if (e && typeof e.preventDefault === 'function') {
+        e.preventDefault();
+      }
+      set(
+        this,
+        'error',
+        typeof data.error.errors !== 'undefined' ?
+          data.error.errors.firstObject : data.error
```

Comment:
We are hard-coding against another "special" shape here, we expect an object with an `error`-property. The template of `DataWriter` also still expects `error.status` and `error.detail` to be available - will this need to be changed in this PR?

May I suggest to allow users to specify a transformation function instead, that will get the error payload passed? That would let the caller specify how the error-payload gets transformed before we set it as the `error`-property on the writer, or have the caller deal with the error entirely:

```hbs
{{!-- allow caller handle or transform the error --}}
<DataWriter @onError={{this.handleError}} />
```
This would allow us to keep the old way as the default behaviour but have callers handle special cases via `@onError`.

Reply:
> We are hard-coding against another "special" shape here

The shape is no longer "special", it's a Javascript event based on an [ErrorEvent](https://developer.mozilla.org/en-US/docs/Web/API/ErrorEvent) which has an error property (containing a standard shaped javascript error)

> May I suggest to allow users to specify a transformation function instead, that will get the error payload passed?

Ideally the component should always stick as closely as possible to existing Javascript objects, it only used/uses non-standard ones because `ember-data`.

Additionally, all of this is dealt with inside the component via interactions between DataSink (the equivalent of DataSource) and DataWriter itself (the equivalent of DataLoader). So as a user you just need to fill out the `error` slot.

> This would allow us to keep the old way as the default behaviour but have callers handle special cases via @onError.

The new way is is now the "default behaviour" the old way is deprecated whilst we switch over to using less `ember-data`, once we finish the switch over there will be no other "cases" apart from the default one. I would prefer this sharp edge whilst we are switching over to be hidden from the user, so inside the component itself in my eyes is the best place to do it - now all instances of DataWriter support standard JS erroring.

---

> still expects error.status and error.detail to be available

Are you referring to this line here?

https://github.com/hashicorp/consul/blob/155c4bc2af77f71ca9cd5067f3c18d1c9ff67ecf/ui/packages/consul-ui/app/components/data-writer/index.hbs#L109-L111

Where we conditionally check if `error.status` and `error.detail` exists?

If so, I guess I can add a check for `error.message` here also, it doesn't effect this PR or the follow up though.

Reply:
aha now I see the actual error that is used here seems to originate from here:

https://github.com/hashicorp/consul/blob/155c4bc2af77f71ca9cd5067f3c18d1c9ff67ecf/ui/packages/consul-ui/app/utils/dom/event-source/index.js#L141-L144

And we can always expect an `ErrorEvent`-like data-structure.

Still think the callers should probably be able to handle this error somehow for flexibility sake but for now I guess it's fine 🤷

Reply:
Here are some docs for that for reference https://developer.mozilla.org/en-US/docs/Web/API/EventSource/error_event

Reply:
Forgot to mention, the value of the `error` property on the `ErrorEvent` does get 'exposed' and you can then use that as flexibly as you like. The error gets yielded out basically and you can use that however you like from thereon.

For example here in policy notifications. We know `Policy` uses ember-data shaped errors so we use that shape here

https://github.com/hashicorp/consul/blob/155c4bc2af77f71ca9cd5067f3c18d1c9ff67ecf/ui/packages/consul-ui/app/components/consul/policy/notifications/index.hbs#L20-L24

I think you mentioned offline that we may want to parse a specific error, so you could do that quite easily here/like this if that was necessary.